<a href="https://colab.research.google.com/github/BHARATJHAWAR52/Hamoye_Spring-2024/blob/main/Hamoye_QuizB_energydata_complete.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [17]:
import pandas as pd
import numpy as np
import seaborn as sns

In [18]:
energy_data=pd.read_csv("/content/drive/MyDrive/HAMOYE SPRING'24/energydata_complete (1).csv")
energy_data.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

In [19]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
X = energy_data[['T2']]
y = energy_data['T6']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15)

model = LinearRegression()
model.fit(X_train, y_train)

# Making predictions on the test set
y_pred = model.predict(X_test)

# Calculating the RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

print("Root Mean Squared Error:", round(rmse, 3))

Root Mean Squared Error: 3.591


In [20]:
energy_data=energy_data.drop(columns=['date','lights'])
energy_data

Appliances         T1       RH_1         T2       RH_2         T3  \
0              60  19.890000  47.596667  19.200000  44.790000  19.790000   
1              60  19.890000  46.693333  19.200000  44.722500  19.790000   
2              50  19.890000  46.300000  19.200000  44.626667  19.790000   
3              50  19.890000  46.066667  19.200000  44.590000  19.790000   
4              60  19.890000  46.333333  19.200000  44.530000  19.790000   
...           ...        ...        ...        ...        ...        ...   
19730         100  25.566667  46.560000  25.890000  42.025714  27.200000   
19731          90  25.500000  46.500000  25.754000  42.080000  27.133333   
19732         270  25.500000  46.596667  25.628571  42.768571  27.050000   
19733         420  25.500000  46.990000  25.414000  43.036000  26.890000   
19734         430  25.500000  46.600000  25.264286  42.971429  26.823333   

            RH_3         T4       RH_4         T5  ...         T9     RH_9  \
0      44.730000  19.000000  45.566667  17.166667  ...  17.033333  45.5300   
1      44.790000  19.000000  45.992500  17.166667  ...  17.066667  45.5600   
2      44.933333  18.926667  45.890000  17.166667  ...  17.000000  45.5000   
3      45.000000  18.890000  45.723333  17.166667  ...  17.000000  45.4000   
4      45.000000  18.890000  45.530000  17.200000  ...  17.000000  45.4000   
...          ...        ...        ...        ...  ...        ...      ...   
19730  41.163333  24.700000  45.590000  23.200000  ...  23.200000  46.7900   
19731  41.223333  24.700000  45.590000  23.230000  ...  23.200000  46.7900   
19732  41.690000  24.700000  45.730000  23.230000  ...  23.200000  46.7900   
19733  41.290000  24.700000  45.790000  23.200000  ...  23.200000  46.8175   
19734  41.156667  24.700000  45.963333  23.200000  ...  23.200000  46.8450   

           T_out  Press_mm_hg     RH_out  Windspeed  Visibility  Tdewpoint  \
0       6.600000        733.5  92.000000   7.000000   63.000000   5.300000   
1       6.483333        733.6  92.000000   6.666667   59.166667   5.200000   
2       6.366667        733.7  92.000000   6.333333   55.333333   5.100000   
3       6.250000        733.8  92.000000   6.000000   51.500000   5.000000   
4       6.133333        733.9  92.000000   5.666667   47.666667   4.900000   
...          ...          ...        ...        ...         ...        ...   
19730  22.733333        755.2  55.666667   3.333333   23.666667  13.333333   
19731  22.600000        755.2  56.000000   3.500000   24.500000  13.300000   
19732  22.466667        755.2  56.333333   3.666667   25.333333  13.266667   
19733  22.333333        755.2  56.666667   3.833333   26.166667  13.233333   
19734  22.200000        755.2  57.000000   4.000000   27.000000  13.200000   

             rv1        rv2  
0      13.275433  13.275433  
1      18.606195  18.606195  
2      28.642668  28.642668  
3      45.410389  45.410389  
4      10.084097  10.084097  
...          ...        ...  
19730  43.096812  43.096812  
19731  49.282940  49.282940  
19732  29.199117  29.199117  
19733   6.322784   6.322784  
19734  34.118851  34.118851  

[19735 rows x 27 columns]

In [21]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
normalised_data= pd.DataFrame(scaler.fit_transform(energy_data),columns=energy_data.columns)
normalised_data

Appliances        T1      RH_1        T2      RH_2        T3      RH_3  \
0        0.046729  0.327350  0.566187  0.225345  0.684038  0.215188  0.746066   
1        0.046729  0.327350  0.541326  0.225345  0.682140  0.215188  0.748871   
2        0.037383  0.327350  0.530502  0.225345  0.679445  0.215188  0.755569   
3        0.037383  0.327350  0.524080  0.225345  0.678414  0.215188  0.758685   
4        0.046729  0.327350  0.531419  0.225345  0.676727  0.215188  0.758685   
...           ...       ...       ...       ...       ...       ...       ...   
19730    0.084112  0.926786  0.537657  0.711655  0.606309  0.830841  0.579374   
19731    0.074766  0.919747  0.536006  0.701769  0.607836  0.825302  0.582178   
19732    0.242991  0.919747  0.538666  0.692651  0.627198  0.818378  0.603988   
19733    0.383178  0.919747  0.549491  0.677054  0.634717  0.805085  0.585294   
19734    0.392523  0.919747  0.538758  0.666171  0.632902  0.799546  0.579062   

             T4      RH_4        T5  ...        T9      RH_9     T_out  \
0      0.351351  0.764262  0.175506  ...  0.223032  0.677290  0.372990   
1      0.351351  0.782437  0.175506  ...  0.226500  0.678532  0.369239   
2      0.344745  0.778062  0.175506  ...  0.219563  0.676049  0.365488   
3      0.341441  0.770949  0.175506  ...  0.219563  0.671909  0.361736   
4      0.341441  0.762697  0.178691  ...  0.219563  0.671909  0.357985   
...         ...       ...       ...  ...       ...       ...       ...   
19730  0.864865  0.765258  0.752031  ...  0.864724  0.729443  0.891747   
19731  0.864865  0.765258  0.754897  ...  0.864724  0.729443  0.887460   
19732  0.864865  0.771233  0.754897  ...  0.864724  0.729443  0.883173   
19733  0.864865  0.773794  0.752031  ...  0.864724  0.730581  0.878885   
19734  0.864865  0.781192  0.752031  ...  0.864724  0.731719  0.874598   

       Press_mm_hg    RH_out  Windspeed  Visibility  Tdewpoint       rv1  \
0         0.097674  0.894737   0.500000    0.953846   0.538462  0.265449   
1         0.100000  0.894737   0.476190    0.894872   0.533937  0.372083   
2         0.102326  0.894737   0.452381    0.835897   0.529412  0.572848   
3         0.104651  0.894737   0.428571    0.776923   0.524887  0.908261   
4         0.106977  0.894737   0.404762    0.717949   0.520362  0.201611   
...            ...       ...        ...         ...        ...       ...   
19730     0.602326  0.416667   0.238095    0.348718   0.901961  0.861981   
19731     0.602326  0.421053   0.250000    0.361538   0.900452  0.985726   
19732     0.602326  0.425439   0.261905    0.374359   0.898944  0.583979   
19733     0.602326  0.429825   0.273810    0.387179   0.897436  0.126371   
19734     0.602326  0.434211   0.285714    0.400000   0.895928  0.682391   

            rv2  
0      0.265449  
1      0.372083  
2      0.572848  
3      0.908261  
4      0.201611  
...         ...  
19730  0.861981  
19731  0.985726  
19732  0.583979  
19733  0.126371  
19734  0.682391  

[19735 rows x 27 columns]

In [36]:
from sklearn.metrics import mean_absolute_error
feature_df=energy_data.drop(columns=['Appliances'])

Appliances=energy_data['Appliances']

x_train, x_test, y_train, y_test = train_test_split(feature_df,Appliances,test_size=0.3,random_state=42)
linear_model = LinearRegression()
linear_model.fit(x_train,y_train)
y_train_predict=linear_model.predict(x_train)
mae=mean_absolute_error(y_train,y_train_predict)
round(mae,3)

53.747

In [37]:
rmse = np.sqrt(mean_squared_error(y_train,y_train_predict))
round(rmse,3)

95.216

In [38]:
y_test_predict=linear_model.predict(x_test)
mae=mean_absolute_error(y_test,y_test_predict)
round(mae,3)

53.647

In [39]:
rmse = np.sqrt(mean_squared_error(y_test,y_test_predict))
round(rmse,3)

93.64

in the training dataset the model has 53.747 mae and 95.216 rmse whereas in the test dataset the model has 53.647 mae and 93.64 rmse. Since the model doesnot show much of a difference in the training and test dataset on the basis of mae and rmse metrices we can say the model does not overfit.


In [42]:
from sklearn.linear_model import Ridge

ridge_model = Ridge()
ridge_model.fit(x_train, y_train)

y_test_pred_ridge = ridge_model.predict(x_test)
rmse_test_ridge = np.sqrt(mean_squared_error(y_test, y_test_pred_ridge))
round(rmse_test_ridge,3)

93.641

In [44]:
from sklearn.linear_model import Lasso

lasso_model = Lasso()
lasso_model.fit(x_test, y_test)

y_test_pred_lasso = lasso_model.predict(x_test)
rmse_test_lasso = np.sqrt(mean_squared_error(y_test, y_test_pred_ridge))
round(rmse_test_lasso,3)

93.641

In [45]:
lasso_model1 = Lasso()
lasso_model1.fit(x_train, y_train)

# Obtain the feature weights
feature_weights = lasso_model1.coef_

# Count the number of non-zero feature weights
non_zero_features = sum(feature_weights != 0)
non_zero_features

21